clar2wasm/src/cost.rs

```diff
@@ -0,0 +1,743 @@
+        let n = n.into();
+
+        self.emit_with_caf(
+            instrs,
+            cost.runtime,
+            self.runtime,
+            ErrorMap::CostOverrunRuntime as _,
+            n,
+        );
+        self.emit_with_caf(
+            instrs,
+            cost.read_count,
+            self.read_count,
+            ErrorMap::CostOverrunReadCount as _,
+            n,
+        );
+        self.emit_with_caf(
+            instrs,
+            cost.read_length,
+            self.read_length,
+            ErrorMap::CostOverrunReadLength as _,
+            n,
+        );
+        self.emit_with_caf(
+            instrs,
+            cost.write_count,
+            self.write_count,
+            ErrorMap::CostOverrunWriteCount as _,
+            n,
+        );
+        self.emit_with_caf(
+            instrs,
+            cost.write_length,
+            self.write_length,
+            ErrorMap::CostOverrunWriteLength as _,
+            n,
+        );
+    }
+
+    fn emit_with_caf(
+        &self,
+        instrs: &mut InstrSeqBuilder,
+        params: Caf,
+        global: GlobalId,
+        err_code: i32,
+        n: impl Into<Scalar>,
+    ) {
+        match params {
+            Caf::Constant(n) => caf_const(instrs, global, self.runtime_error, err_code, n),
+            Caf::Linear { a, b } => {
+                caf_linear(instrs, global, self.runtime_error, err_code, n, a, b)
+            }
+            Caf::NLogN { a, b } => caf_nlogn(instrs, global, self.runtime_error, err_code, n, a, b),
+            Caf::None => {}
+        }
+    }
+}
+
+// NOTE: since we're only measuring costs during runtime - and everything
+//       else is left for the analyzer
+
+fn caf_const(
+    instrs: &mut InstrSeqBuilder,
+    global: GlobalId,
+    error: FunctionId,
+    err_code: i32,
+    cost: impl Into<Scalar>,
+) {
+    let cost = cost.into();
+
+    instrs.global_get(global).scalar_get(cost);
+
+    // global - cost
+    instrs
+        .instr(Instr::Binop(Binop {
+            op: BinaryOp::I64Sub,
+        }))
+        .global_set(global)
+        .global_get(global)
+        .i64_const(0)
+        .instr(Instr::Binop(Binop {
+            op: BinaryOp::I64LtS,
+        }))
+        .if_else(
+            None,
+            |builder| {
+                builder.i32_const(err_code);
+                builder.call(error);
+            },
+            |_| {},
+        );
+}
+
+fn caf_linear(
+    instrs: &mut InstrSeqBuilder,
+    global: GlobalId,
+    error: FunctionId,
+    err_code: i32,
+    n: impl Into<Scalar>,
+    a: u64,
+    b: u64,
+) {
+    let n = n.into();
+
+    // cost = (+ (* a n) b))
+    instrs
+        .global_get(global)
+        .i64_const(b as _)
+        .i64_const(a as _)
+        .scalar_get(n)
+        // * a
+        .instr(Instr::Binop(Binop {
+            op: BinaryOp::I64Mul,
+        }))
+        // + b
+        .instr(Instr::Binop(Binop {
+            op: BinaryOp::I64Add,
+        }));
```

Comment:
It could be clearer if you did things after their corresponding comments:
```rust
// n
.scalar_get(n)
// * a
.i64_const(a)
.binop(BinaryOp::I64Mul)
// + b
.i64_const(b)
.binop(BinaryOp::i64Add)
```
Btw, [this exists](https://docs.rs/walrus/latest/walrus/struct.InstrSeqBuilder.html#method.binop).

Reply:
https://github.com/stacks-network/clarity-wasm/pull/622#discussion_r2021214350